### DISCUSSION TITLE
API calls inside getInitialProps even it's static build <= not 100% static site??

### DISCUSSION THREAD

**Discussion by codyng** [vercel/next.js#10879] [Help]
Hi there,

In my front page, I have made query data from CMS and passed to the front page:

Below is the piece of code in my exportPathMap function:

```
        const res = await fetch(SOME_ENDPOINT);
        const data = await res.json();
        const paths = {
            '/': { page: '/', query: { data  } }
        };
```

And in my Index.getInitialProps() function, I did the same API call and return it as the initial props:

```
Index.getInitialProps = async (ctx) => {
        const res = await fetch(SOME_ENDPOINT);
        const data = await res.json();
        return { data }
}
```

Coz I have followed the tutorial from here: Fetching Data for Pages
https://nextjs.org/learn/basics/fetching-data-for-pages

I've checked the server request logs and found that during the page is rendered in server side, the API will be requested. It's fine. When jumping to other page and going back to the front page (using &lt;Link&gt; not &lt;a&gt;), the Index.getInitialProps() gets called again and so does the API endpoint call.

Is this the normal usage? I think it's not 100% static. Suppose the API end point cannot be accessed for the public. It must be used during the build and export phase.

The index.html must be generated in early time and content should be generated to meet the SEO purpose. So what is the recommended and correct usage in my case?

Thanks!!!!!

**Comment by timneutkens** [comment]
See https://github.com/zeit/next.js/issues/9524

**Comment by NeftaliYagua** [comment]
# 📋 TODO List - Cumplimiento OWASP Top 10

**Última actualización:** 2026-02-05  
**Orden:** De más fácil a más difícil

---

## ✅ Completadas

- [x] Implementar protección CSRF/XSRF completa (A05, A08)
- [x] Eliminar dependencias no utilizadas (@google-cloud/storage, fast-xml-parser)
- [x] Actualizar dependencias vulnerables críticas (Angular, Firebase Admin)

---

## 🔵 Fácil (1-2 horas)

### 1. Documentar política de rotación de secrets
- **Categoría:** A02 - Cryptographic Failures
- **Descripción:** Crear documentación sobre cuándo y cómo rotar secrets (API keys, tokens, credenciales)
- **Archivos:** `docs/security/secrets-rotation.md`
- **Esfuerzo:** 30 minutos

### 2. Agregar validación explícita de ownership en endpoints críticos
- **Categoría:** A01 - Broken Access Control
- **Descripción:** Revisar endpoints que modifican recursos y agregar validación explícita de ownership
- **Archivos:** `functions/routes/admin/*.js`, `functions/routes/trainer/*.js`
- **Esfuerzo:** 1-2 horas

### 3. Configurar Content Security Policy (CSP) básico
- **Categoría:** A05 - Security Misconfiguration
- **Descripción:** Agregar headers CSP básicos en Express
- **Archivos:** `functions/server.js`
- **Esfuerzo:** 1 hora
- **Nota:** Ya existe estructura, solo falta completar configuración

---

## 🟡 Medio (2-4 horas)

### 4. Implementar rate limiting básico por IP
- **Categoría:** A01 - Broken Access Control
- **Descripción:** Instalar y configurar `express-rate-limit` para protección básica contra DoS
- **Archivos:** `functions/server.js`, `functions/middleware/rate-limit.js` (nuevo)
- **Dependencias:** `express-rate-limit`
- **Esfuerzo:** 2-3 horas
- **Prioridad:** Alta (mencionado en análisis como pendiente)

### 5. Mejorar logging estructurado
- **Categoría:** A09 - Security Logging and Monitoring Failures
- **Descripción:** Implementar logging estructurado con niveles, contexto y formato JSON
- **Archivos:** `functions/utils/logger.js` (nuevo), `functions/server.js`
- **Dependencias:** `winston` o `pino`
- **Esfuerzo:** 3-4 horas

### 6. Implementar verificación de integridad de dependencias
- **Categoría:** A08 - Software and Data Integrity Failures
- **Descripción:** Configurar `npm audit` en CI/CD y verificación de checksums
- **Archivos:** `.github/workflows/ci.yml`, `package.json` (scripts)
- **Esfuerzo:** 2 horas

### 7. Configurar CSP completo y avanzado
- **Categoría:** A05 - Security Misconfiguration
- **Descripción:** Configurar CSP completo con nonces, report-uri, y políticas específicas por ruta
- **Archivos:** `functions/middleware/csp.js` (nuevo), `functions/server.js`
- **Esfuerzo:** 3-4 horas

---

## 🟠 Difícil (4-8 horas)

### 8. Implementar rate limiting avanzado por usuario/IP
- **Categoría:** A01 - Broken Access Control
- **Descripción:** Rate limiting diferenciado por usuario autenticado vs IP, con almacenamiento en Redis
- **Archivos:** `functions/middleware/rate-limit.js`, `functions/server.js`
- **Dependencias:** `express-rate-limit`, `ioredis` (ya instalado)
- **Esfuerzo:** 4-6 horas
- **Prioridad:** Alta

### 9. Implementar monitoreo de seguridad y alertas
- **Categoría:** A09 - Security Logging and Monitoring Failures
- **Descripción:** Configurar alertas automáticas para eventos de seguridad (intentos de acceso fallidos, rate limit excedido, etc.)
- **Archivos:** `functions/utils/security-monitor.js` (nuevo)
- **Integraciones:** Firebase Cloud Logging, webhooks
- **Esfuerzo:** 6-8 horas

### 10. Auditoría completa de dependencias y actualización
- **Categoría:** A06 - Vulnerable and Outdated Components
- **Descripción:** Revisar todas las dependencias transitivas, actualizar las vulnerables, y documentar decisiones
- **Archivos:** `package.json`, `functions/package.json`, documentación
- **Esfuerzo:** 4-6 horas
- **Nota:** Actualmente hay 2 vulnerabilidades transitivas pendientes

### 11. Implementar validación de entrada más robusta con esquemas
- **Categoría:** A03 - Injection, A04 - Insecure Design
- **Descripción:** Implementar validación de esquemas centralizada (ej: Joi, Zod) para todos los endpoints
- **Archivos:** `functions/middleware/validation.js` (nuevo), todos los routers
- **Dependencias:** `joi` o `zod`
- **Esfuerzo:** 6-8 horas

---

## 🔴 Muy Difícil (8+ horas)

### 12. Implementar WAF (Web Application Firewall) o protección avanzada
- **Categoría:** A05 - Security Misconfiguration, A01 - Broken Access Control
- **Descripción:** Configurar reglas avanzadas de protección (detección de patrones de ataque, bloqueo automático)
- **Archivos:** Configuración de infraestructura, middleware personalizado
- **Integraciones:** Cloudflare, AWS WAF, o solución propia
- **Esfuerzo:** 8-12 horas

### 13. Implementar análisis estático de seguridad (SAST)
- **Categoría:** A06 - Vulnerable and Outdated Components, A04 - Insecure Design
- **Descripción:** Integrar herramientas de análisis estático (SonarQube, Snyk, etc.) en CI/CD
- **Archivos:** `.github/workflows/security-scan.yml` (nuevo), configuración de herramientas
- **Esfuerzo:** 8-10 horas

### 14. Implementar pruebas de seguridad automatizadas (DAST)
- **Categoría:** A04 - Insecure Design, A05 - Security Misconfiguration
- **Descripción:** Integrar herramientas de análisis dinámico (OWASP ZAP, Burp Suite) en CI/CD
- **Archivos:** `.github/workflows/security-scan.yml`, configuración de herramientas
- **Esfuerzo:** 10-12 horas

### 15. Revisión completa de arquitectura de seguridad
- **Categoría:** A04 - Insecure Design
- **Descripción:** Revisión arquitectónica completa con enfoque en seguridad, threat modeling, y diseño seguro
- **Archivos:** Documentación de arquitectura, diagramas de seguridad
- **Esfuerzo:** 12-16 horas

---

## 📊 Métricas de Progreso

- **Completadas:** 3/15 (20%)
- **Fácil:** 0/3 (0%)
- **Medio:** 0/7 (0%)
- **Difícil:** 0/3 (0%)
- **Muy Difícil:** 0/2 (0%)

---

## 🎯 Prioridades Recomendadas

1. **Inmediato:** #4 (Rate limiting básico) - Protección crítica contra DoS
2. **Corto plazo:** #1, #2, #3 - Mejoras rápidas de seguridad
3. **Medio plazo:** #8, #9, #10 - Mejoras significativas de seguridad
4. **Largo plazo:** #12, #13, #14, #15 - Mejoras avanzadas y arquitectónicas

---

## 📝 Notas

- Las vulnerabilidades transitivas restantes (2) se resolverán cuando `firebase-admin` actualice sus dependencias
- El rate limiting es crítico y debe implementarse pronto
- La documentación de seguridad debe mantenerse actualizada con cada cambio
